### PR TITLE
Limit number of runs processed from RunListProgressiveRendering

### PR DESCRIPTION
**Pull request by res0nance**
We were having an issue where build history on a particular node was causing huge GC spikes. Protect against this by limiting the number of runs processed.

### Proposed changelog entries

* Entry 1: Limit number of items processed in build history page to 100

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by fqueiruga — COMMENTED** [review]
If the build history list on the widget is truncated, I'd say we need a "view all builds" hyperlink at the bottom of the list that takes the user to the trends page.

File: core/src/main/java/jenkins/widgets/RunListProgressiveRendering.java [hunk 1]
```diff
@@ -66,8 +61,11 @@ public void setBuilds(Iterable<? extends Run<?,?>> builds) {
             synchronized (this) {
                 results.add(element);
             }
-            decay *= (1 - 1 / MAX_LIKELY_RUNS);
-            progress(1 - decay);
+            progress(runs / MAX_RUNS);
```
Comment (rsandell) [inline core/src/main/java/jenkins/widgets/RunListProgressiveRendering.java:64]: If the number of actual runs are less than 100 the progress will never get to 100%

Comment (res0nance) [reply core/src/main/java/jenkins/widgets/RunListProgressiveRendering.java:64]: I think the progress disappears once fully loaded

**Comment by res0nance** [comment]
> If the build history list on the widget is truncated, I'd say we need a "view all builds" hyperlink at the bottom of the list that takes the user to the trends page.

I wasn't thinking of this from a UX perspective. I'm not sure which page you mean by trends page. 

The reason for this PR is in some scenarios this code can cause poor performance as it loops through unbounded number of builds.

**Comment by fqueiruga** [comment]
I mean the page linked by this:
Captura de pantalla 2020-04-23 a las 13 04 49

I think users should not lose the ability to check all builds, and to at least see at the bottom of the list that there are hidden items.

**Comment by res0nance** [comment]
> I mean the page linked by this:
> Captura de pantalla 2020-04-23 a las 13 04 49
> 
> I think users should not lose the ability to check all builds, and to at least see at the bottom of the list that there are hidden items.

Thats, true I'll close this and work on this

**Comment by daniel-beck** [comment]
The build history referenced here is the one for specific nodes. It can be extremely expensive to compute; job specific pages don't get anywhere close. It certainly makes sense to limit it, perhaps even to remove it entirely in its current form.

**Comment by res0nance** [comment]
> The build history referenced here is the one for specific nodes. It can be extremely expensive to compute; job specific pages don't get anywhere close. It certainly makes sense to limit it, perhaps even to remove it entirely in its current form.

Yes my issue is regarding node specific builds. But from what @fqueiruga says this does affect the History widget for the job page. Not sure what is the best way to limit this? Perhaps some additional logic to add a limit and enforce it from jelly?

**Pull request closed without merging**